File: decibel-meter/android/app/src/main/java/com/izowooi/dm/core/LevelEstimate.kt
package com.izowooi.dm.core

/** A display estimate, separate from the PCM measurement and reference calibration. */
data class LevelEstimate(val offset: Double, val basis: String) {
    fun decibels(dbfs: Double): Double {
        if (!dbfs.isFinite() || !offset.isFinite() || dbfs <= Acoustics.FLOOR_DB) return 0.0
        return (dbfs + offset).coerceAtLeast(0.0)
    }

    companion object {
        fun resolve(input: InputConditions?, calibration: CalibrationProfile?): LevelEstimate {
            if (input != null && calibration?.validFor(input) == true) {
                return LevelEstimate(calibration.offset, "reference_calibration")
            }
            // Android CDD nominal sensitivities for the built-in microphone, not device calibration.
            // UNPROCESSED: 94 dB SPL -> -36 dBFS; VOICE_RECOGNITION: 90 -> -22.35.
            if (input?.routeId?.endsWith(":15") == true) {
                if (input.source.startsWith("UNPROCESSED;")) return LevelEstimate(130.0, "android_unprocessed_nominal_v1")
                if (input.source.startsWith("VOICE_RECOGNITION;")) return LevelEstimate(112.35, "android_voice_nominal_v1")
            }
            return LevelEstimate(110.0, "generic_estimate_v1")
        }
    }
}

File: decibel-meter/ios/dm/LevelEstimate.swift
import Foundation

/// A display estimate, separate from the PCM measurement and reference calibration.
struct LevelEstimate: Codable, Equatable, Sendable {
    let offset: Double
    let basis: String

    static func resolve(input: InputConditions?, calibration: CalibrationProfile?) -> LevelEstimate {
        if let input, let calibration, calibration.valid(for: input) {
            return LevelEstimate(offset: calibration.offset, basis: "reference_calibration")
        }
        // A common, deliberately approximate starting point; not an iPhone model calibration.
        return LevelEstimate(offset: 110, basis: "generic_estimate_v1")
    }

    func decibels(_ dbfs: Double) -> Double {
        guard dbfs.isFinite, offset.isFinite else { return 0 }
        // Digital silence is never presented as a positive ambient sound level.
        guard dbfs > Acoustics.floorDB else { return 0 }
        return max(0, dbfs + offset)
    }
}
